bob-apps/bin/bobt.rs

```diff
@@ -0,0 +1,434 @@
+        log::info!("GET: count: {}, latency: {}", self.get_count, latency);
+        let latency = self.put_time.as_secs_f64() / self.put_count as f64 * 1000.0;
+        log::info!("PUT: count: {}, latency: {}", self.put_count, latency);
+        let latency = self.delete_time.as_secs_f64() / self.delete_count as f64 * 1000.0;
+        log::info!("DELETE: count: {}, latency: {}", self.delete_count, latency);
+    }
+
+    async fn put(&mut self, key: u64, size: usize) -> Result<(), String> {
+        let data = vec![1; size];
+        let mut url = Url::parse(&self.settings.api_uri.to_string()).unwrap();
+        url.set_path(&format!("/data/{}", self.settings.convert_key(key)));
+        let addr = url.to_string();
+        log::info!("Put to {}", addr);
+        let sw = Stopwatch::new();
+        let req = self.http_client
+            .post(addr)
+            .body(data)
+            .header("username", "admin")
+            .header("password", "password")
+            .build()
+            .map_err(|e| e.to_string())?;
+        let res = self.http_client.execute(req).await.map_err(|e| e.to_string())?;
+        if !res.status().is_success() {
+            log::warn!("Put resulted in error code: {:?}", res.status());
+        }
+        log::info!(
+            "Put {} with size {}",
+            key,
+            size,
+        );
+        self.put_time += sw.elapsed();
+        self.put_count += 1;
+        Ok(())
+    }
+
+    async fn get(&mut self, key: u64) -> Result<usize, String> {
+        let mut url = Url::parse(&self.settings.api_uri.to_string()).unwrap();
+        url.set_path(&format!("/data/{}", self.settings.convert_key(key)));
+        let addr = url.to_string();//format!("{}/data/{}", self.settings.api_uri, self.settings.convert_key(key));
+        let req = self.http_client
+            .get(addr)
+            .header("username", "admin")
+            .header("password", "password")
+            .build()
+            .map_err(|e| e.to_string())?;
+        let sw = Stopwatch::new();
+        let res = self.http_client.execute(req).await.map_err(|e| e.to_string())?;
+        let status = res.status();
+        if !status.is_success() {
+            if status != StatusCode::NOT_FOUND {
+                log::warn!("Get resulted in error code: {:?}", res.status());
+            }
+            return Err(status.to_string());
+        }
+        self.get_time += sw.elapsed();
+        self.get_count += 1;
+        log::debug!("Get {} result: {:?}", key, res);
+        let bytes = res.bytes().await;
+        bytes.map(|b| b.len()).map_err(|e| e.to_string())
+    }
+
+    async fn delete(&mut self, key: u64) -> Result<usize, String> {
+        let size = self.get(key).await?;
+        log::debug!("Delete {} with size {}", key, size);
+        let req = self
+            .http_client
+            .delete(format!("{}/data/{}", self.settings.api_uri, self.settings.convert_key(key)))
```

Comment:
It would be good to use the same approach to build URL as used in `put` and `get`